pkg/datasource.go

```diff
@@ -118,8 +176,15 @@ func (o *OCIDatasource) QueryData(ctx context.Context, req *backend.QueryDataReq
 
 	queryType := ts.QueryType
 
-	if len(o.tenancyAccess) == 0 || ts.TenancyMode == "multitenancy" {
-		err := o.getConfigProvider(ts.Environment, ts.TenancyMode)
+	o.logger.Debug("QueryType")
+	o.logger.Debug(ts.QueryType)
+	o.logger.Debug(ts.Environment)
+	o.logger.Debug(ts.Tenancy)
+	o.logger.Debug(ts.TenancyMode)
+
+	// if len(o.tenancyAccess) == 0 || ts.TenancyMode == "multitenancy" {
+	if len(o.tenancyAccess) == 0 {
+		err := o.getConfigProvider(ts.Environment, ts.TenancyMode, req)
```

Comment:
Hi. Many thanks for your directions here. I changed the code accordingly. How does it look like now ?

Reply:
Looks like you've done what I suggested in the other comment. Great 👍 

> Question: Can you change environment/tenancymode on a per-query basis? If not would strongly suggest calling getConfigProvider in NewOCIDatasource. Or basically, can you populate o.tenancyAccess when datasource has been configured without having to take special measures depending on what query it is? Doing this will simplify a lot of things and you would be guaranteed by the grafana-plugin-sdk-go that the NewOCIDatasource is only executed by one goroutine at a time and by that removing the need for synchronization/locks as I referred to in my earlier comment.